File: 백준/Gold/15684. 사다리 조작/사다리 조작.swift
let nmh = readLine()!.split{$0 == " "}.compactMap{Int(String($0))}
let (n,m,h) = (nmh[0],nmh[1],nmh[2])

var v = Array(repeating: Array(repeating: false, count: n), count: h)

for _ in stride(from: 0, to: m, by: 1) {
  let input = readLine()!.split{$0 == " "}.compactMap{Int(String($0))}
  let (h,i) = (input[0], input[1])
  v[h-1][i-1] = true
}

// // i to i 판단 로직
func check() -> Bool {
  for i in stride(from: 0, to: n, by: 1) {
    var start = i
    
    for j in stride(from: 0, to: h, by: 1) {
      if v[j][start] { start += 1 }                     // 우
      else if start > 0 && v[j][start-1] { start -= 1 } // 좌
    }

    if start != i {
      return false
    }
  }
  return true
}

func go(_ cnt: Int, _ x: Int, _ y: Int) {
  if check() {
    ret = min(ret, cnt)
    return
  }

  if cnt == 3 || cnt >= ret {
    return
  }

  // 상하좌우 이동하면서 사다리 놓을 수 있는지 체크
  for i in stride(from: x, to: h, by: 1) {
    let cur = i == x ? y : 0
    if cur > n-1 { continue }

    for j in stride(from: cur, to: n-1, by: 1) {
      // 이미 놓여있으면 현재 사다리 가로로 못 놓음
      if j > 0 && v[i][j-1] { continue }
      if v[i][j] || v[i][j+1] { continue }

      v[i][j] = true
      go(cnt + 1, i, j+2)
      v[i][j] = false
    }
  }
}

var ret = 4
go(0,0,0)
print(ret < 4 ? ret : -1)

File: 백준/Gold/15684. 사다리 조작/사다리 조작.cc
#include <bits/stdc++.h>
using namespace std;
const int INF = 987654321;
int n, m, h, a, b, ret = INF, visited[34][34];
bool check(){
    for(int i = 1; i <= n; i++){
        int start = i;
        for(int j = 1; j <= h; j++){
            if(visited[j][start])start++;
            else if(visited[j][start - 1])start--;
        }
        if(start != i) return false;
    }
    return true;
}
void go(int here, int cnt){ 
    if(cnt > 3 || cnt >= ret) return;
    if(check()){
        ret = min(ret, cnt); return;
    }
    for(int i = here; i <= h; i++){
        for(int j = 1; j <= n; j++){
            if(visited[i][j] || visited[i][j - 1] || visited[i][j + 1]) continue;
            visited[i][j] = 1;
            go(i, cnt + 1);
            visited[i][j] = 0;
        }
    }
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
    cout.tie(NULL);
    cin >> n >> m >> h;
    for(int i = 0; i < m; i++){
        cin >> a >> b; visited[a][b] = 1;
    }
    go(1, 0);
    cout << ((ret == INF) ? -1 : ret) << "\n";
	return 0;
}